drivers/sdhc/sdhc_esp32.c

```diff
@@ -0,0 +1,1544 @@
+	uint32_t timeout_cfg;
+	int ret_esp;
+	int ret = 0;
+
+	/* convert command structures Zephyr vs ESP */
+	struct sdmmc_command esp_cmd = {
+		.opcode = cmd->opcode,
+		.arg = cmd->arg,
+	};
+
+	if (data) {
+		esp_cmd.data = data->data;
+		esp_cmd.blklen = data->block_size;
+		esp_cmd.datalen = (data->blocks * data->block_size);
+		esp_cmd.buflen = esp_cmd.datalen;
+		timeout_cfg = data->timeout_ms;
+	} else {
+		timeout_cfg = cmd->timeout_ms;
+	}
+
+	/* setting timeout according to command type */
+	if (cmd->timeout_ms == SDHC_TIMEOUT_FOREVER) {
+		esp_cmd.timeout_ms = SDMMC_TIMEOUT_MAX;
+	} else {
+		esp_cmd.timeout_ms = timeout_cfg;
+	}
+
+	/*
+	 * Determine type of command. Note that driver is expected to
+	 * handle CMD12 and CMD23 for reading and writing blocks
+	 */
+	switch (cmd->opcode) {
+	case SD_GO_IDLE_STATE:
+		esp_cmd.flags = SCF_CMD_BC | SCF_RSP_R0;
+		break;
+
+	case SD_APP_CMD:
+	case SD_SEND_STATUS:
+	case SD_SET_BLOCK_SIZE:
+		esp_cmd.flags = SCF_CMD_AC | SCF_RSP_R1;
+		break;
+
+	case SD_SEND_IF_COND:
+		esp_cmd.flags = SCF_CMD_BCR | SCF_RSP_R7;
+		break;
+
+	case SD_APP_SEND_OP_COND:
+		esp_cmd.flags = SCF_CMD_BCR | SCF_RSP_R3;
+
+#define SD_OCR_SDHC_CAP (1 << 30)
+#define SD_OCR_VOL_MASK 0xFF8000 /* bits 23:15 */
+		esp_cmd.arg = SD_OCR_SDHC_CAP | SD_OCR_VOL_MASK;
+		break;
+
+	case SDIO_RW_DIRECT:
+		esp_cmd.flags = SCF_CMD_AC | SCF_RSP_R5;
+		break;
+
+	case SDIO_SEND_OP_COND:
+		esp_cmd.flags = SCF_CMD_BCR | SCF_RSP_R4;
+		break;
+
+	case SD_ALL_SEND_CID:
+		esp_cmd.flags = SCF_CMD_BCR | SCF_RSP_R2;
+		break;
+
+	case SD_SEND_RELATIVE_ADDR:
+		esp_cmd.flags = SCF_CMD_BCR | SCF_RSP_R6;
+		break;
+
+	case SD_SEND_CSD:
+		esp_cmd.flags = SCF_CMD_AC | SCF_RSP_R2;
+		esp_cmd.datalen = 0;
+		break;
+
+	case SD_SELECT_CARD:
+		/* Don't expect to see a response when de-selecting a card */
+		esp_cmd.flags = SCF_CMD_AC | (cmd->arg > 0 ? SCF_RSP_R1 : 0);
+		break;
+
+	case SD_APP_SEND_SCR:
+	case SD_SWITCH:
+	case SD_READ_SINGLE_BLOCK:
+	case SD_READ_MULTIPLE_BLOCK:
+	case SD_APP_SEND_NUM_WRITTEN_BLK:
+		esp_cmd.flags = SCF_CMD_ADTC | SCF_CMD_READ | SCF_RSP_R1;
+		break;
+
+	case SD_WRITE_SINGLE_BLOCK:
+	case SD_WRITE_MULTIPLE_BLOCK:
+		esp_cmd.flags = SCF_CMD_ADTC | SCF_RSP_R1;
+		break;
+
+	default:
+		LOG_INF("SDHC driver: command %u not supported", cmd->opcode);
+		return -ENOTSUP;
+	}
+
+	while (retries > 0) {
+
+		ret_esp = sdmmc_host_do_transaction(cfg->slot, &esp_cmd);
+
+		if (ret_esp) {
+			retries--; /* error, retry */
+		} else {
+			break;
+		}
+	}
+
+	if ((ret_esp != 0) || esp_cmd.error) {
+		LOG_DBG("\nError for command: %u arg %08x ret_esp = 0x%x error = 0x%x\n",
+			cmd->opcode, cmd->arg, ret_esp, esp_cmd.error);
+
+		ret_esp = (ret_esp > 0) ? ret_esp : esp_cmd.error;
+
+		ret = err_esp2zep(ret_esp);
+	} else {
+		/* fill response buffer */
+		memcpy(cmd->response, esp_cmd.response, sizeof(cmd->response));
+
+#define MMC_R1_CURRENT_STATE(resp) (((resp)[0] >> 9) & 0xf)
```

Comment:
Please provide this `define` outside of the function scope (ideally near the top of the file with other defines), perhaps with a comment describing what it does